bftclient/src/bft_client.cpp

```diff
@@ -0,0 +1,165 @@
+// subcomponent's license, as noted in the LICENSE file.
+
+#include "bftclient/bft_client.h"
+#include "bftengine/ClientMsgs.hpp"
+#include "assertUtils.hpp"
+
+namespace bft::client {
+
+// This function creates a ClientRequestMsg or a ClientPreProcessRequestMsg depending upon config.
+//
+// Since both of these are just instancds of a `ClientRequestMsgHeader` followed by the message
+// data, we construct them here, rather than relying on the type constructors embedded into the
+// bftEngine impl. This allows us to not have to link with the bftengine library, and also allows us
+// to return the messages as vectors with proper RAII based memory management.
+Msg MakeClientMsg(const RequestConfig& config, Msg&& request, bool read_only, uint16_t client_id) {
+  uint8_t flags = read_only ? READ_ONLY_REQ : EMPTY_FLAGS_REQ;
+  if (config.pre_execute) {
+    flags |= PRE_PROCESS_REQ;
+  }
+
+  auto header_size = sizeof(bftEngine::ClientRequestMsgHeader);
+
+  Msg msg(header_size + request.size());
+  bftEngine::ClientRequestMsgHeader* header = reinterpret_cast<bftEngine::ClientRequestMsgHeader*>(msg.data());
+  header->msgType = config.pre_execute ? PRE_PROCESS_REQUEST_MSG_TYPE : REQUEST_MSG_TYPE;
+  header->spanContextSize = config.span_context.size();
+  header->idOfClientProxy = client_id;
+  header->flags = flags;
+  header->reqSeqNum = config.sequence_number;
+  header->requestLength = msg.size();
+  header->timeoutMilli = config.timeout.count();
+  header->cid_length = config.correlation_id.size();
+
+  auto* position = msg.data() + header_size;
+
+  // Copy the span context
+  std::memcpy(position, config.span_context.data(), config.span_context.size());
+  position += config.span_context.size();
+
+  // Copy the request data
+  std::memcpy(position, request.data(), request.size());
+  position += request.size();
+
+  // Copy the correlation ID
+  std::memcpy(position, config.correlation_id.data(), config.correlation_id.size());
+
+  return msg;
+}
+
+Reply Client::Send(const WriteConfig& config, Msg&& request) {
+  Assert(!outstanding_request_.has_value());
+  auto match_config = WriteConfigToMatchConfig(config);
+  bool read_only = false;
+  return Send(match_config, config.request, std::move(request), read_only);
+}
+
+Reply Client::Send(const ReadConfig& config, Msg&& request) {
+  Assert(!outstanding_request_.has_value());
+  auto match_config = ReadConfigToMatchConfig(config);
+  bool read_only = true;
+  return Send(match_config, config.request, std::move(request), read_only);
+}
+
+Reply Client::Send(const MatchConfig& match_config,
+                   const RequestConfig& request_config,
+                   Msg&& request,
+                   bool read_only) {
+  outstanding_request_ = Matcher(match_config);
+  receiver_.activate(request_config.max_reply_size);
+  const auto& msg = MakeClientMsg(request_config, std::move(request), read_only, config_.id.val);
```

Comment:
Nit: Might be a bit confusing to bind to a reference, could just do `const auto msg`.

Reply:
Since I've proposed to use the reference, I feel like, I need to answer this:)
Using the const reference guarantees that the bound' object lifetime is extended to the reference's lifetime.
Thus, we use the object by aliasing w/o copying or moving it.
While on the contrary, if `const auto` is used either copy or move happens.
The behavior is guaranteed by the standard since C++98(if I am not mistaken).

Reply:
Oh, I haven't seen that you've discussed that, sorry.

You are right about lifetime extension. I think the reference will bind to a temporary, meaning that without RVO there will still be a move. And with RVO, there won't be any moves/copies. Having said that, I don't think ref/without reference will have any impact on the assembly. My comment was mainly focused on code style, i.e. when I see a reference, I wouldn't expect that this function returns a new object. Rather, I expect it returns a message that's already constructed. What I mean is I prefer to keep it simple in that case :) In general, not a big deal, we can leave as is.

In terms of assembly in O3, you can see they look identical: https://godbolt.org/z/qGF25B

Reply:
To be more precise, C++17's rules are a bit different from the C++14 ones, but I think the result is the same: https://godbolt.org/z/qD5-N8

https://en.cppreference.com/w/cpp/language/copy_elision

I am almost sure of @andrewjstone 's reaction and face expression when reading this... :)

Reply:
@dartdart26 , Interesting comparison!
Nit: If I am not mistaken, this case is not `RVO` but `NRVO`, where copy-elision is not guaranteed by the standard.

Reply:
> Having said that, I don't think ref/without reference will have any impact on the assembly. My comment was mainly focused on code style, i.e. when I see a reference, I wouldn't expect that this function returns a new object. Rather, I expect it returns a message that's already constructed. What I mean is I prefer to keep it simple in that case :) In general, not a big deal, we can leave as is.

I actually agree with this sentiment. After reflecting, I think the `const auto&` makes the code less clear. I think I'm going to change it back in the next commit.